### PR TITLE
feat(rollup-relayer & gas-oracle): graceful restart

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

Doc: https://www.notion.so/scrollzkp/Graceful-restart-of-rollup-relayer-and-gas-oracle-cf194c09c06c44f7a2215f7d4a36a4d8

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: rollup/internal/orm/transaction.go
Comment (georgehao) [inline rollup/internal/orm/transaction.go]: Seems don't need this

File: database/migrate/migrations/00015_transaction.sql
Comment (georgehao) [inline database/migrate/migrations/00015_transaction.sql:19]: how about use `TIMESTAMP`

Comment (georgehao) [inline database/migrate/migrations/00015_transaction.sql]: this is confused, in general `xx_at` is a timestamp. So I think can change to `submit_block_number`

Comment (georgehao) [reply database/migrate/migrations/00015_transaction.sql]: and reserve the submit_at to record the submit time

File: rollup/internal/controller/sender/sender.go
Comment (georgehao) [inline rollup/internal/controller/sender/sender.go]: If deployed replicas, I think this number is not right.

if receipt on A, but insert on B, the inserted block number will be wrong.

File: rollup/internal/orm/transaction.go
Comment (georgehao) [inline rollup/internal/orm/transaction.go]: I think there need to check the status to prevent repeated update

File: rollup/internal/orm/transaction.go
Comment (colinlyguo) [reply rollup/internal/orm/transaction.go]: Need to use `RLPEncoding: rlp.Bytes(),` to store the content of the transaction.

Comment (georgehao) [reply rollup/internal/orm/transaction.go]: oh, i misread the var

File: database/migrate/migrations/00015_transaction.sql
Comment (georgehao) [inline database/migrate/migrations/00015_transaction.sql]: how about change the name to `pending_transaction`?

Comment (colinlyguo) [reply database/migrate/migrations/00015_transaction.sql]: changed in [88de67b](https://github.com/scroll-tech/scroll/pull/1076/commits/88de67b28bfa9a0183d642c1d6e44a37ffefb732)

File: database/migrate/migrations/00015_transaction.sql
Comment (colinlyguo) [reply database/migrate/migrations/00015_transaction.sql]: changed in [88de67b](https://github.com/scroll-tech/scroll/pull/1076/commits/88de67b28bfa9a0183d642c1d6e44a37ffefb732).
and in inserting a new pending transaction, `submit_at` is equal to `created_at`, so we can check the submit time in `created_at`.

File: rollup/internal/orm/transaction.go
Comment (colinlyguo) [reply rollup/internal/orm/transaction.go]: Yes, added in [88de67b](https://github.com/scroll-tech/scroll/pull/1076/commits/88de67b28bfa9a0183d642c1d6e44a37ffefb732).

File: rollup/internal/controller/sender/sender.go
Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go]: Yes. `submit_block_number` is not the block number for which the transaction is included. It only reflects the roughly submit time so that the sender can decide whether to resubmit or not.

Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go]: If the node's sync heights of each replica are the same (or up to the latest block), it's OK.

But if not, there would be some problem triggering the resubmit logic, either too early or too late.

Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go]: fixed in [f22b884](https://github.com/scroll-tech/scroll/pull/1076/commits/f22b884667637d4fd164dca65ea6df6a12c71d76)

File: rollup/internal/controller/sender/sender.go
Comment (zimpha) [inline rollup/internal/controller/sender/sender.go:470]: one possible error is that: `one of the old transactions is confirmed` and in such case we should mark the transaction confirmed. But it rarely happens, we can leave it manually.

See the comments of the original function for more context.

Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:470]: Yes. I use the following logic to handle this case in the loop:
1. If the tx is` confirmed`, update current tx's status as `confirmed`, and the sender's other txs (of the same `nonce`) as `failed`.
https://github.com/scroll-tech/scroll/blob/442c296ec5e3c5866bbdecc8df53bf9f2ee5d00a/rollup/internal/controller/sender/sender.go#L419-L435
2. When resubmitting txs, check if the "pending" tx has been marked `failed` because it may be confirmed by other "replaced" txs within the same loop.
https://github.com/scroll-tech/scroll/blob/442c296ec5e3c5866bbdecc8df53bf9f2ee5d00a/rollup/internal/controller/sender/sender.go#L449-L457

Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:467]: seems that the above logic still has some problems... I need to separate the loop into **`two loops`**, one for confirm checking, and the other for resubmitting.

Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:470]: added some unit tests covering this case in [1dd1d64](https://github.com/scroll-tech/scroll/pull/1076/commits/1dd1d64a9eaa376d44e3f79369f9c3cd88f92f67)

**Review by zimpha — DISMISSED** [review]

**Review by georgehao — DISMISSED** [review]

**Comment by colinlyguo** [comment]
still testing `access list` feature. don't merge this PR yet.

**Comment by colinlyguo** [comment]
TODOs after merging this PR:
- [x] sepolia: db migrate up and redeployment.
- [x] sepolia: testing graceful restart.
- [x] sepolia: checking db info correctness and CPU usage.
- [ ] grafana: reconfigured metrics.
- [ ] mainnet: db migrate up and redeployment.

File: common/types/db.go
Comment (0xmountaintop) [inline common/types/db.go]:
```suggestion
		return fmt.Sprintf("Unknown SenderType (%d)", int32(t))
```

File: common/types/db.go
Comment (0xmountaintop) [inline common/types/db.go]:
```suggestion
		return fmt.Sprintf("Unknown TxStatus (%d)", int32(s))
```

File: common/types/db_test.go
Comment (colinlyguo) [inline common/types/db_test.go]:
```suggestion
			"Unknown SenderType (999)",
```

File: common/types/db_test.go
Comment (colinlyguo) [inline common/types/db_test.go]:
```suggestion
			"Unknown TxStatus (999)",
```

**Review by georgehao — APPROVED** [review]

**Review by zimpha — APPROVED** [review]

**Pull request merged**
